File: apps/macos/Packages/DailyDoListDomain/Sources/DailyDoListDomain/Markdown/TaskTracker.swift
import DailyDoListModels
import Foundation

/// A task with a stable identity. Identity survives edits, reordering and status changes so that
/// agent threads stay attached to "the same" to-do item while the user rewrites it.
public struct TrackedTask: Hashable, Sendable, Codable, Identifiable {
  public var id: String
  public var text: String
  public var status: TaskStatus
  public var line: Int
  public var depth: Int
  public var parentId: String?
  public var notes: [String]
  /// Epoch ms when the task was first observed.
  public var firstSeenAt: EpochMillis
  /// Epoch ms of the last text/notes/status change.
  public var updatedAt: EpochMillis
  /// The agent wrote this task (see `ParsedTask.agent`).
  public var agent: Bool

  public init(
    id: String, text: String, status: TaskStatus, line: Int, depth: Int, parentId: String?,
    notes: [String], firstSeenAt: EpochMillis, updatedAt: EpochMillis, agent: Bool = false
  ) {
    self.id = id
    self.text = text
    self.status = status
    self.line = line
    self.depth = depth
    self.parentId = parentId
    self.notes = notes
    self.firstSeenAt = firstSeenAt
    self.updatedAt = updatedAt
    self.agent = agent
  }
}

public enum TaskChangeKind: String, Sendable, Hashable, Codable {
  case text, notes
}

/// What changed between two tracking passes.
public struct TaskDiff: Hashable, Sendable {
  public struct Update: Hashable, Sendable {
    public var task: TrackedTask
    public var previous: TrackedTask
    /// Text and/or notes, in that order.
    public var changes: [TaskChangeKind]
  }

  public struct StatusChange: Hashable, Sendable {
    public var task: TrackedTask
    public var previous: TrackedTask
  }

  /// New tasks, in document order.
  public var added: [TrackedTask] = []
  /// Tasks whose text and/or notes changed, in document order.
  public var updated: [Update] = []
  public var statusChanged: [StatusChange] = []
  /// Previous tasks that are gone, in their previous order.
  public var removed: [TrackedTask] = []

  public init() {}

  /// `isEmptyDiff`.
  public var isEmpty: Bool {
    added.isEmpty && updated.isEmpty && statusChanged.isEmpty && removed.isEmpty
  }
}

public struct TaskTrackingResult: Hashable, Sendable {
  /// One tracked task per parsed task, in document order.
  public var tasks: [TrackedTask]
  public var diff: TaskDiff
}

/// Task identity across edits, port of @ddl/core `markdown/task-tracker.ts`.
public enum TaskTracker {
  public static let defaultSimilarityThreshold = 0.5

  /// `trackTasks`: matches freshly parsed tasks against previously tracked ones. Passes, in order:
  /// 1. identical normalized text; a text found once on each side is an anchor, duplicates are
  ///    aligned in order, preferring the lines the anchors predict;
  /// 2. fuzzy: prefix extension (still typing, counts as 0.9) or Dice similarity ≥ threshold,
  ///    minus 0.01 per line moved (at most 0.2), assigned greedily by best score;
  /// 3. same line with weak similarity (≥ 0.3): an in-place rewrite.
  /// Unmatched parsed tasks get ids from `idFactory` (called in document order) and are `added`;
  /// unmatched previous tasks are `removed`.
  public static func track(
    previous: [TrackedTask],
    parsed: [ParsedTask],
    now: EpochMillis = currentMillis(),
    similarityThreshold: Double = defaultSimilarityThreshold,
    idFactory: () -> String = { TaskTracker.makeID() }
  ) -> TaskTrackingResult {
    let matchOf = TaskMatcher(
      previous: previous.map { ($0.text, $0.line) }, parsed: parsed.map { ($0.text, $0.line) },
      threshold: similarityThreshold
    ).match()

    var tasks: [TrackedTask] = []
    tasks.reserveCapacity(parsed.count)
    var usedPrevious = [Bool](repeating: false, count: previous.count)
    // Per parsed task: nil when added, else which of text / notes / status changed.
    var changes: [(text: Bool, notes: Bool, status: Bool)?] = []
    changes.reserveCapacity(parsed.count)
    for (pi, p) in parsed.enumerated() {
      let ci = matchOf[pi]
      guard ci >= 0 else {
        tasks.append(
          TrackedTask(
            id: idFactory(), text: p.text, status: p.status, line: p.line, depth: p.depth,
            parentId: nil,
            notes: p.notes, firstSeenAt: now, updatedAt: now, agent: p.agent))
        changes.append(nil)
        continue
      }
      let prev = previous[ci]
      usedPrevious[ci] = true
      let change = (
        text: !prev.text.jsEquals(p.text), notes: !sameNotes(prev.notes, p.notes),
        status: prev.status != p.status
      )
      tasks.append(
        TrackedTask(
          id: prev.id, text: p.text, status: p.status, line: p.line, depth: p.depth, parentId: nil,
          notes: p.notes, firstSeenAt: prev.firstSeenAt,
          updatedAt: change.text || change.notes || change.status ? now : prev.updatedAt,
          agent: p.agent))
      changes.append(change)
    }
    let taskAtLine = lineLookup(parsed)
    for (pi, p) in parsed.enumerated() {
      if let line = p.parentLine, let parent = taskAtLine(line) {
        tasks[pi].parentId = tasks[parent].id
      }
    }

    var diff = TaskDiff()
    for (pi, task) in tasks.enumerated() {
      guard let change = changes[pi] else {
        diff.added.append(task)
        continue
      }
      let prev = previous[matchOf[pi]]
      var kinds: [TaskChangeKind] = []
      if change.text { kinds.append(.text) }
      if change.notes { kinds.append(.notes) }
      if !kinds.isEmpty { diff.updated.append(.init(task: task, previous: prev, changes: kinds)) }
      if change.status { diff.statusChanged.append(.init(task: task, previous: prev)) }
    }
    for (ci, prev) in previous.enumerated() where !usedPrevious[ci] { diff.removed.append(prev) }
    return TaskTrackingResult(tasks: tasks, diff: diff)
  }

  /// The index of the last parsed task on a line (the core's `idByLine` map): a binary search
  /// when lines increase, as they do coming from the parser.
  private static func lineLookup(_ parsed: [ParsedTask]) -> (Int) -> Int? {
    let lines = parsed.map(\.line)
    if zip(lines, lines.dropFirst()).allSatisfy({ $0 < $1 }) {
      return { line in
        let index = TaskMatcher.lowerBound(lines, line)
        return index < lines.count && lines[index] == line ? index : nil
      }
    }
    var byLine: [Int: Int] = [:]
    for (index, line) in lines.enumerated() { byLine[line] = index }
    return { byLine[$0] }
  }

  /// `track` on a document's text.
  public static func track(
    previous: [TrackedTask], markdown: String, now: EpochMillis = currentMillis(),
    similarityThreshold: Double = defaultSimilarityThreshold,
    idFactory: () -> String = { TaskTracker.makeID() }
  ) -> TaskTrackingResult {
    track(
      previous: previous, parsed: TaskParser.parse(markdown), now: now,
      similarityThreshold: similarityThreshold, idFactory: idFactory)
  }

  /// The core's default id: `tsk_` and 10 random base-36 characters.
  public static func makeID() -> String {
    let alphabet = Array("0123456789abcdefghijklmnopqrstuvwxyz".utf8)
    var generator = SystemRandomNumberGenerator()
    let body = (0..<10).map { _ in alphabet[Int(UInt8.random(in: 0...255, using: &generator)) % 36]
    }
    return "tsk_" + String(decoding: body, as: UTF8.self)
  }

  /// Whole milliseconds since the epoch, like `Date.now()`.
  public static func currentMillis() -> EpochMillis {
    (Date().timeIntervalSince1970 * 1000).rounded(.down)
  }

  private static func sameNotes(_ a: [String], _ b: [String]) -> Bool {
    a.count == b.count && zip(a, b).allSatisfy { $0.jsEquals($1) }
  }
}

/// The three matching passes of `trackTasks`.
struct TaskMatcher {
  /// Above this many candidate pairs, the fuzzy pass only compares tasks on nearby lines.
  static let fuzzyPairBudget = 40_000
  static let minFuzzyWindow = 20
  /// Duplicate groups bigger than this (parsed × previous) are aligned greedily, not optimally.
  static let alignCellBudget = 250_000
  /// Weight of the displacement cost over the raw line distance used to break its ties.
  static let tieScale = 1 << 20

  let previousLines: [Int]
  let parsedLines: [Int]
  let previousText: [TextFeatures]
  let parsedText: [TextFeatures]
  let threshold: Double

  init(
    previous: [(text: String, line: Int)], parsed: [(text: String, line: Int)], threshold: Double
  ) {
    previousLines = previous.map(\.line)
    parsedLines = parsed.map(\.line)
    previousText = previous.map { TextFeatures($0.text) }
    parsedText = parsed.map { TextFeatures($0.text) }
    self.threshold = threshold
  }

  /// Group numbers of identical normalized texts, numbered in first-seen order (open addressing
  /// on the text hash; a flat table stays fast in unoptimized builds, unlike `Dictionary`).
  static func groups(_ features: [TextFeatures]) -> (groupOf: [Int], count: Int) {
    var capacity = 16
    while capacity < features.count * 2 { capacity <<= 1 }
    let mask = capacity - 1
    var slots = [Int](repeating: -1, count: capacity)  // index of the group's first text
    var groupOf = [Int](repeating: 0, count: features.count)
    var count = 0
    slots.withUnsafeMutableBufferPointer { slotBuffer in
      groupOf.withUnsafeMutableBufferPointer { groupBuffer in
        let s = slotBuffer.baseAddress!
        for (index, text) in features.enumerated() {
          var slot =
            Int(
              truncatingIfNeeded: UInt64(bitPattern: Int64(text.hash)) &* 0x9E37_79B9_7F4A_7C15
                >> 32) & mask
          while true {
            let first = s[slot]
            if first < 0 {
              s[slot] = index
              groupBuffer[index] = count
              count += 1
              break
            }
            if features[first].isEqual(to: text) {
              groupBuffer[index] = groupBuffer[first]
              break
            }
            slot = (slot + 1) & mask
          }
        }
      }
    }
    return (groupOf, count)
  }

  /// For each parsed task, the index of the previous task it continues, or -1.
  func match() -> [Int] {
    var matchOf = [Int](repeating: -1, count: parsedLines.count)
    var usedPrevious = [Bool](repeating: false, count: previousLines.count)
    func link(_ pi: Int, _ ci: Int) {
      matchOf[pi] = ci
      usedPrevious[ci] = true
    }

    // Pass 1: identical normalized text, grouped in first-seen order (previous tasks first).
    let (groupOf, groupCount) = Self.groups(previousText + parsedText)
    var groupPrevious = [[Int]](repeating: [], count: groupCount)
    var groupParsed = [[Int]](repeating: [], count: groupCount)
    for ci in previousText.indices { groupPrevious[groupOf[ci]].append(ci) }
    for pi in parsedText.indices { groupParsed[groupOf[previousText.count + pi]].append(pi) }

    var anchors: [(line: Int, shift: Int, order: Int)] = []
    for g in groupPrevious.indices where groupPrevious[g].count == 1 && groupParsed[g].count == 1 {
      let pi = groupParsed[g][0]
      let ci = groupPrevious[g][0]
      link(pi, ci)
      anchors.append((parsedLines[pi], parsedLines[pi] - previousLines[ci], anchors.count))
    }
    anchors.sort { ($0.line, $0.order) < ($1.line, $1.order) }
    /// Where a parsed line sat before the edit, judging by the nearest anchor (above first).
    func expectedLine(_ line: Int) -> Int {
      var lo = 0
      var hi = anchors.count
      while lo < hi {
        let mid = (lo + hi) >> 1
        if anchors[mid].line <= line { lo = mid + 1 } else { hi = mid }
      }
      let anchor = lo > 0 ? anchors[lo - 1] : anchors.first
      return line - (anchor?.shift ?? 0)
    }

    var expected = [Int?](repeating: nil, count: parsedLines.count)
    func duplicateCost(_ pi: Int, _ ci: Int) -> Double {
      let p = parsedLines[pi]
      let c = previousLines[ci]
      let e: Int
      if let cached = expected[pi] {
        e = cached
      } else {
        e = expectedLine(p)
        expected[pi] = e
      }
      return Double(abs(e - c)) * Double(Self.tieScale) + Double(min(abs(p - c), Self.tieScale - 1))
    }
    for g in groupPrevious.indices {
      let prev = groupPrevious[g]
      let next = groupParsed[g]
      if prev.isEmpty || next.isEmpty || (prev.count == 1 && next.count == 1) { continue }
      Self.alignDuplicates(
        next: next.sorted { (parsedLines[$0], $0) < (parsedLines[$1], $1) },
        previous: prev.sorted { (previousLines[$0], $0) < (previousLines[$1], $1) },
        cost: duplicateCost, link: link)
    }

    // Pass 2: fuzzy matching among the leftovers.
    let openParsed = parsedLines.indices.filter { matchOf[$0] < 0 }
    let openPrevious = previousLines.indices.filter { !usedPrevious[$0] }
    var pairs: [(pi: Int, ci: Int, score: Double, order: Int)] = []
    func consider(_ pi: Int, _ ci: Int) {
      let a = parsedText[pi]
      let b = previousText[ci]
      let extends = a.extends(b)
      // Dice can't beat its bound: skipping those pairs changes nothing but saves the bigrams.
      if !extends, a.count >= 2, b.count >= 2, a.similarityBound(b) < threshold { return }
      var similarity = a.similarity(b)
      if extends { similarity = max(similarity, 0.9) }
      if similarity < threshold { return }
      let penalty = Double(min(abs(previousLines[ci] - parsedLines[pi]), 20)) * 0.01
      pairs.append((pi, ci, similarity - penalty, pairs.count))
    }
    if openParsed.count * openPrevious.count <= Self.fuzzyPairBudget {
      for pi in openParsed {
        for ci in openPrevious { consider(pi, ci) }
      }
    } else {
      // A mass rewrite: only nearby lines can plausibly be the same task.
      let window = max(Self.minFuzzyWindow, Self.fuzzyPairBudget / (2 * openParsed.count))
      let sorted = openPrevious.sorted { (previousLines[$0], $0) < (previousLines[$1], $1) }
      let lines = sorted.map { previousLines[$0] }
      for pi in openParsed {
        let center = expectedLine(parsedLines[pi])
        var k = Self.lowerBound(lines, center - window)
        while k < sorted.count && lines[k] <= center + window {
          consider(pi, sorted[k])
          k += 1
        }
      }
    }
    pairs.sort { $0.score != $1.score ? $0.score > $1.score : $0.order < $1.order }
    for pair in pairs where matchOf[pair.pi] < 0 && !usedPrevious[pair.ci] {
      link(pair.pi, pair.ci)
    }

    // Pass 3: in-place rewrite on the same line.
    var openByLine: [Int: [Int]] = [:]
    for ci in previousLines.indices where !usedPrevious[ci] {
      openByLine[previousLines[ci], default: []].append(ci)
    }
    for pi in parsedLines.indices where matchOf[pi] < 0 {
      guard let ci = openByLine[parsedLines[pi]]?.first(where: { !usedPrevious[$0] }) else {
        continue
      }
      if parsedText[pi].similarity(previousText[ci]) >= 0.3 { link(pi, ci) }
    }
    return matchOf
  }

  /// Pairs up tasks that share one text. Both lists are sorted by line; every task of the shorter
  /// list is matched, in order, to the task of the longer list that minimizes the total cost.
  static func alignDuplicates(
    next: [Int], previous: [Int], cost: (Int, Int) -> Double, link: (Int, Int) -> Void
  ) {
    let nextIsLong = next.count >= previous.count
    let long = nextIsLong ? next : previous
    let short = nextIsLong ? previous : next
    func pairCost(_ l: Int, _ s: Int) -> Double { nextIsLong ? cost(l, s) : cost(s, l) }
    func pair(_ l: Int, _ s: Int) { nextIsLong ? link(l, s) : link(s, l) }

    if long.count * short.count > alignCellBudget {
      var taken = [Bool](repeating: false, count: long.count)
      for s in short {
        var best = -1
        var bestCost = 0.0
        for i in long.indices where !taken[i] {
          let c = pairCost(long[i], s)
          if best == -1 || c < bestCost {
            best = i
            bestCost = c
          }
        }
        taken[best] = true
        pair(long[best], s)
      }
      return
    }

    // best[i][j]: cheapest way to match short[0..<j] to an increasing subsequence of long[0..<i].
    let cols = short.count + 1
    var best = [Double](repeating: .infinity, count: (long.count + 1) * cols)
    var took = [Bool](repeating: false, count: (long.count + 1) * cols)
    best.withUnsafeMutableBufferPointer { bestBuffer in
      took.withUnsafeMutableBufferPointer { tookBuffer in
        let b = bestBuffer.baseAddress!
        let t = tookBuffer.baseAddress!
        for i in 0...long.count { b[i * cols] = 0 }
        guard !long.isEmpty, !short.isEmpty else { return }
        for i in 1...long.count {
          for j in 1...min(i, short.count) {
            let skip = b[(i - 1) * cols + j]
            let take = b[(i - 1) * cols + j - 1] + pairCost(long[i - 1], short[j - 1])
            // Ties keep the earlier task of the longer list, like a nearest-line scan would.
            if take < skip {
              b[i * cols + j] = take
              t[i * cols + j] = true
            } else {
              b[i * cols + j] = skip
            }
          }
        }
      }
    }
    var i = long.count
    var j = short.count
    while j > 0 {
      if took[i * cols + j] {
        pair(long[i - 1], short[j - 1])
        j -= 1
      }
      i -= 1
    }
  }

  static func lowerBound(_ sorted: [Int], _ value: Int) -> Int {
    var lo = 0
    var hi = sorted.count
    while lo < hi {
      let mid = (lo + hi) >> 1
      if sorted[mid] < value { lo = mid + 1 } else { hi = mid }
    }
    return lo
  }
}

File: packages/core/src/markdown/task-tracker.ts
import { createId } from "../ids";
import { normalizeText } from "../text";
import type { ParsedTask, TaskStatus } from "./tasks";

/**
 * A task with a stable identity. Identity survives edits, reordering and status changes so that
 * agent threads stay attached to "the same" to-do item while the user rewrites it.
 */
export interface TrackedTask {
  id: string;
  text: string;
  status: TaskStatus;
  line: number;
  depth: number;
  parentId: string | null;
  notes: string[];
  /** Epoch ms when the task was first observed. */
  firstSeenAt: number;
  /** Epoch ms of the last text/notes/status change. */
  updatedAt: number;
}

export type TaskChangeKind = "text" | "notes";

export interface TaskDiff {
  added: TrackedTask[];
  /** Text and/or notes changed. */
  updated: Array<{ task: TrackedTask; previous: TrackedTask; changes: TaskChangeKind[] }>;
  statusChanged: Array<{ task: TrackedTask; previous: TrackedTask }>;
  removed: TrackedTask[];
}

export interface TrackOptions {
  now?: number;
  idFactory?: () => string;
  /** Minimum fuzzy similarity (0..1) to treat an edited line as the same task. Default 0.5. */
  similarityThreshold?: number;
}

/** Above this many candidate pairs, the fuzzy pass only compares tasks on nearby lines. */
const FUZZY_PAIR_BUDGET = 40_000;
const MIN_FUZZY_WINDOW = 20;
/** Duplicate groups bigger than this (parsed × previous) are aligned greedily, not optimally. */
const ALIGN_CELL_BUDGET = 250_000;
/** Weight of the displacement cost over the raw line distance used to break its ties. */
const TIE_SCALE = 1 << 20;

export function isEmptyDiff(diff: TaskDiff): boolean {
  return (
    diff.added.length === 0 &&
    diff.updated.length === 0 &&
    diff.statusChanged.length === 0 &&
    diff.removed.length === 0
  );
}

function sameNotes(a: readonly string[], b: readonly string[]): boolean {
  return a.length === b.length && a.every((n, i) => n === b[i]);
}

/**
 * Matches freshly parsed tasks against previously tracked ones and returns the new tracked list
 * plus a diff. Matching passes, in order:
 *  1. identical normalized text. A text found once on each side is an anchor; duplicates are
 *     aligned in order, preferring the lines the anchors predict (so deleting, inserting or typing
 *     one of several identical tasks leaves the others' ids alone);
 *  2. fuzzy: prefix-extension (still typing) or Dice similarity ≥ threshold, with a small
 *     line-distance penalty, assigned greedily by best score;
 *  3. same line index with weak similarity (≥ 0.3) — an in-place rewrite.
 * Unmatched parsed tasks are `added`; unmatched previous tasks are `removed`.
 */
export function trackTasks(
  previous: readonly TrackedTask[],
  parsed: readonly ParsedTask[],
  options: TrackOptions = {},
): { tasks: TrackedTask[]; diff: TaskDiff } {
  const now = options.now ?? Date.now();
  const idFactory = options.idFactory ?? (() => createId("tsk", 10));
  const matchOf = matchTasks(previous, parsed, options.similarityThreshold ?? 0.5);

  const diff: TaskDiff = { added: [], updated: [], statusChanged: [], removed: [] };
  const usedPrev = new Array<boolean>(previous.length).fill(false);
  const idByLine = new Map<number, string>();
  const tasks: TrackedTask[] = parsed.map((p, pi) => {
    const prevIndex = matchOf[pi]!;
    const prev = prevIndex === -1 ? undefined : previous[prevIndex];
    if (prev) usedPrev[prevIndex] = true;
    const id = prev?.id ?? idFactory();
    idByLine.set(p.line, id);
    const task: TrackedTask = {
      id,
      text: p.text,
      status: p.status,
      line: p.line,
      depth: p.depth,
      parentId: null, // resolved below once all ids are known
      notes: [...p.notes],
      firstSeenAt: prev?.firstSeenAt ?? now,
      updatedAt: now,
    };
    if (!prev) {
      diff.added.push(task);
      return task;
    }
    const changes: TaskChangeKind[] = [];
    if (prev.text !== p.text) changes.push("text");
    if (!sameNotes(prev.notes, p.notes)) changes.push("notes");
    const statusChanged = prev.status !== p.status;
    if (changes.length > 0) diff.updated.push({ task, previous: prev, changes });
    if (statusChanged) diff.statusChanged.push({ task, previous: prev });
    if (changes.length === 0 && !statusChanged) task.updatedAt = prev.updatedAt;
    return task;
  });

  parsed.forEach((p, pi) => {
    tasks[pi]!.parentId = p.parentLine === null ? null : (idByLine.get(p.parentLine) ?? null);
  });
  previous.forEach((t, ci) => {
    if (!usedPrev[ci]) diff.removed.push(t);
  });

  return { tasks, diff };
}

interface LineText {
  text: string;
  line: number;
}

/** For each parsed task, the index of the previous task it continues, or -1. */
function matchTasks(
  previous: readonly LineText[],
  parsed: readonly LineText[],
  threshold: number,
): number[] {
  const matchOf = new Array<number>(parsed.length).fill(-1);
  const usedPrev = new Array<boolean>(previous.length).fill(false);
  const link = (pi: number, ci: number) => {
    matchOf[pi] = ci;
    usedPrev[ci] = true;
  };
  const prevText = previous.map((t) => new TextFeatures(t.text));
  const nextText = parsed.map((p) => new TextFeatures(p.text));

  // Pass 1: identical normalized text.
  const groups = new Map<string, { prev: number[]; next: number[] }>();
  const groupOf = (key: string) => {
    let group = groups.get(key);
    if (!group) {
      group = { prev: [], next: [] };
      groups.set(key, group);
    }
    return group;
  };
  for (const [ci, f] of prevText.entries()) groupOf(f.normalized).prev.push(ci);
  for (const [pi, f] of nextText.entries()) groupOf(f.normalized).next.push(pi);
  const anchors: Array<{ line: number; shift: number }> = [];
  for (const { prev, next } of groups.values()) {
    if (prev.length !== 1 || next.length !== 1) continue;
    link(next[0]!, prev[0]!);
    const line = parsed[next[0]!]!.line;
    anchors.push({ line, shift: line - previous[prev[0]!]!.line });
  }
  anchors.sort((a, b) => a.line - b.line);
  /** Where a parsed line sat before the edit, judging by the nearest anchor (above first). */
  const expectedLine = (line: number) => {
    let lo = 0;
    let hi = anchors.length;
    while (lo < hi) {
      const mid = (lo + hi) >> 1;
      if (anchors[mid]!.line <= line) lo = mid + 1;
      else hi = mid;
    }
    const anchor = anchors[lo - 1] ?? anchors[lo];
    return line - (anchor?.shift ?? 0);
  };
  const expected = new Float64Array(parsed.length).fill(Number.NaN);
  const duplicateCost = (pi: number, ci: number) => {
    const p = parsed[pi]!.line;
    const c = previous[ci]!.line;
    if (Number.isNaN(expected[pi]!)) expected[pi] = expectedLine(p);
    return Math.abs(expected[pi]! - c) * TIE_SCALE + Math.min(Math.abs(p - c), TIE_SCALE - 1);
  };
  for (const { prev, next } of groups.values()) {
    if (prev.length === 0 || next.length === 0 || (prev.length === 1 && next.length === 1)) {
      continue;
    }
    const byLine = (list: readonly LineText[]) => (a: number, b: number) =>
      list[a]!.line - list[b]!.line || a - b;
    alignDuplicates(
      [...next].sort(byLine(parsed)),
      [...prev].sort(byLine(previous)),
      duplicateCost,
      link,
    );
  }

  // Pass 2: fuzzy matching among the leftovers.
  const openNext = parsed.flatMap((_, pi) => (matchOf[pi] === -1 ? [pi] : []));
  const openPrev = previous.flatMap((_, ci) => (usedPrev[ci] ? [] : [ci]));
  const pairs: Array<{ pi: number; ci: number; score: number }> = [];
  const consider = (pi: number, ci: number) => {
    let sim = nextText[pi]!.similarity(prevText[ci]!);
    if (nextText[pi]!.extends(prevText[ci]!)) sim = Math.max(sim, 0.9);
    if (sim < threshold) return;
    const distancePenalty = Math.min(Math.abs(previous[ci]!.line - parsed[pi]!.line), 20) * 0.01;
    pairs.push({ pi, ci, score: sim - distancePenalty });
  };
  if (openNext.length * openPrev.length <= FUZZY_PAIR_BUDGET) {
    for (const pi of openNext) for (const ci of openPrev) consider(pi, ci);
  } else {
    // A mass rewrite: only nearby lines can plausibly be the same task, and comparing everything
    // with everything would block the event loop for seconds on a long note.
    const window = Math.max(
      MIN_FUZZY_WINDOW,
      Math.floor(FUZZY_PAIR_BUDGET / (2 * openNext.length)),
    );
    const sorted = [...openPrev].sort((a, b) => previous[a]!.line - previous[b]!.line || a - b);
    const lines = sorted.map((ci) => previous[ci]!.line);
    for (const pi of openNext) {
      const center = expectedLine(parsed[pi]!.line);
      for (let k = lowerBound(lines, center - window); k < sorted.length; k++) {
        if (lines[k]! > center + window) break;
        consider(pi, sorted[k]!);
      }
    }
  }
  pairs.sort((a, b) => b.score - a.score);
  for (const { pi, ci } of pairs) {
    if (matchOf[pi] !== -1 || usedPrev[ci]) continue;
    link(pi, ci);
  }

  // Pass 3: in-place rewrite on the same line.
  const openByLine = new Map<number, number[]>();
  previous.forEach((t, ci) => {
    if (usedPrev[ci]) return;
    const list = openByLine.get(t.line);
    if (list) list.push(ci);
    else openByLine.set(t.line, [ci]);
  });
  parsed.forEach((p, pi) => {
    if (matchOf[pi] !== -1) return;
    const ci = openByLine.get(p.line)?.find((i) => !usedPrev[i]);
    if (ci !== undefined && nextText[pi]!.similarity(prevText[ci]!) >= 0.3) link(pi, ci);
  });

  return matchOf;
}

/**
 * Pairs up tasks that share one text. Both lists are sorted by line; every task of the shorter
 * list is matched, in order, to the task of the longer list that minimizes the total cost.
 */
function alignDuplicates(
  next: readonly number[],
  prev: readonly number[],
  cost: (pi: number, ci: number) => number,
  link: (pi: number, ci: number) => void,
): void {
  const nextIsLong = next.length >= prev.length;
  const long = nextIsLong ? next : prev;
  const short = nextIsLong ? prev : next;
  const pairCost = (l: number, s: number) => (nextIsLong ? cost(l, s) : cost(s, l));
  const pair = (l: number, s: number) => (nextIsLong ? link(l, s) : link(s, l));

  if (long.length * short.length > ALIGN_CELL_BUDGET) {
    const taken = new Array<boolean>(long.length).fill(false);
    for (const s of short) {
      let best = -1;
      for (let i = 0; i < long.length; i++) {
        if (!taken[i] && (best === -1 || pairCost(long[i]!, s) < pairCost(long[best]!, s))) {
          best = i;
        }
      }
      taken[best] = true;
      pair(long[best]!, s);
    }
    return;
  }

  // best[i][j]: cheapest way to match short[0..j) to an increasing subsequence of long[0..i).
  const cols = short.length + 1;
  const best = new Float64Array((long.length + 1) * cols).fill(Number.POSITIVE_INFINITY);
  const took = new Uint8Array((long.length + 1) * cols);
  for (let i = 0; i <= long.length; i++) best[i * cols] = 0;
  for (let i = 1; i <= long.length; i++) {
    for (let j = 1; j <= Math.min(i, short.length); j++) {
      const skip = best[(i - 1) * cols + j]!;
      const take = best[(i - 1) * cols + j - 1]! + pairCost(long[i - 1]!, short[j - 1]!);
      // Ties keep the earlier task of the longer list, like a nearest-line scan would.
      if (take < skip) {
        best[i * cols + j] = take;
        took[i * cols + j] = 1;
      } else {
        best[i * cols + j] = skip;
      }
    }
  }
  for (let i = long.length, j = short.length; j > 0; i--) {
    if (took[i * cols + j]) {
      pair(long[i - 1]!, short[j - 1]!);
      j--;
    }
  }
}

function lowerBound(sorted: readonly number[], value: number): number {
  let lo = 0;
  let hi = sorted.length;
  while (lo < hi) {
    const mid = (lo + hi) >> 1;
    if (sorted[mid]! < value) lo = mid + 1;
    else hi = mid;
  }
  return lo;
}

/** A task text, normalized once, with lazily built bigram counts for Dice similarity. */
class TextFeatures {
  readonly normalized: string;
  private counts: Map<string, number> | undefined;

  constructor(text: string) {
    this.normalized = normalizeText(text);
  }

  /** Same value as `diceSimilarity` on the original texts. */
  similarity(other: TextFeatures): number {
    const x = this.normalized;
    const y = other.normalized;
    if (x === y) return 1;
    if (x.length < 2 || y.length < 2) return 0;
    const [small, large] =
      x.length <= y.length ? [this.bigrams(), other.bigrams()] : [other.bigrams(), this.bigrams()];
    let overlap = 0;
    for (const [bigram, count] of small) overlap += Math.min(count, large.get(bigram) ?? 0);
    return (2 * overlap) / (x.length - 1 + (y.length - 1));
  }

  /** Same value as `isPrefixExtension` on the original texts. */
  extends(other: TextFeatures, minLength = 3): boolean {
    const x = this.normalized;
    const y = other.normalized;
    if (x.length < minLength || y.length < minLength) return false;
    return x.startsWith(y) || y.startsWith(x);
  }

  private bigrams(): Map<string, number> {
    if (!this.counts) {
      this.counts = new Map();
      const s = this.normalized;
      for (let i = 0; i < s.length - 1; i++) {
        const bigram = s.slice(i, i + 2);
        this.counts.set(bigram, (this.counts.get(bigram) ?? 0) + 1);
      }
    }
    return this.counts;
  }
}
